dags/exchange_rate/clickhouse.py

```diff
@@ -0,0 +1,88 @@
+from typing import Any
+
+import dagster
+
+from .exchange_rate import fetch_exchange_rates
+from clickhouse_driver import Client
+from posthog.clickhouse.cluster import ClickhouseCluster, NodeRole
+
+
+# TODO: Use the constants from `posthog/models/exchange_rate/sql.py` once that's merged
+# rather than hardcoding the table/dict name here
+@dagster.asset(
+    partitions_def=dagster.DailyPartitionsDefinition(start_date="2025-01-01"),
+    group_name="exchange_rates",
+    ins={"exchange_rates_data": dagster.AssetIn(key=fetch_exchange_rates.key)},
+)
+def store_exchange_rates_in_clickhouse(
+    context: dagster.AssetExecutionContext,
+    exchange_rates_data: dict[str, Any],
+    cluster: dagster.ResourceParam[ClickhouseCluster],
+) -> dagster.MaterializeResult:
+    """
+    Stores exchange rates data in ClickHouse.
+
+    This asset depends on the fetch_exchange_rates asset and will be executed
+    after it completes successfully.
+
+    The base currency is always USD as per the table design.
+    """
+    # Extract data from the input
+    date_str = context.partition_key
+
+    # Transform data into rows for ClickHouse
+    rows = [{"date": date_str, "currency": currency, "rate": rate} for currency, rate in exchange_rates_data.items()]
+
+    # Log information about the data being stored
+    context.log.info(f"Storing {len(rows)} exchange rates for {date_str} in ClickHouse")
+
+    # Prepare values for batch insert
+    # Use toDate() to cast the string date to a ClickHouse Date type
+    values = [f"(toDate('{row['date']}'), '{row['currency']}', {row['rate']})" for row in rows]
+
+    # Execute the insert if there are values to insert
+    insert_sql = ""
+    if values:
+        insert_sql = f"""
+        INSERT INTO exchange_rate (date, currency, rate)
+        VALUES {', '.join(values)}
+        """
+
+        def insert(client: Client):
+            try:
+                # Batch insert all values
+                client.sync_execute(insert_sql)
+                context.log.info("Successfully inserted exchange rates")
+            except Exception as e:
+                context.log.warning(f"Failed to insert exchange rates: {e}")
+
+        def reload_dict(client: Client):
+            try:
+                client.sync_execute("SYSTEM RELOAD DICTIONARY exchange_rate_dict")
+                context.log.info("Successfully reloaded exchange_rate_dict dictionary")
+            except Exception as e:
+                context.log.warning(f"Failed to reload exchange_rate_dict dictionary: {e}")
+
+        cluster.map_hosts_by_role(insert, NodeRole.DATA).result()
+        cluster.map_hosts_by_role(reload_dict, NodeRole.DATA).result()
```

Comment:
@PostHog/clickhouse is this reasonable? Or should I be running in different places? I need to insert data and then once all data has been inserted I need to reload the dictionaries